libpod/runtime_img.go

```diff
@@ -171,21 +254,36 @@ func (r *Runtime) PullImage(imgName string, allTags bool, signaturePolicyPath st
 	}
 	defer policyContext.Destroy()
 
-	copyOptions := common.GetCopyOptions(reportWriter, signaturePolicyPath, nil, nil, common.SigningOptions{})
+	copyOptions := common.GetCopyOptions(reportWriter, signaturePolicyPath, nil, &options.DockerRegistryOptions, options.SigningOptions)
+
 	for _, image := range images {
+		fmt.Printf("Trying to pull %s...", image)
 		reference := image
-		if srcRef.DockerReference() != nil {
-			reference = srcRef.DockerReference().String()
+		var pullRef types.ImageReference
+		if fromRegistry {
+			// need to get ImageReference
+			pullRef, err = alltransports.ParseImageName(image)
+			if err != nil {
+				return errors.Errorf("error parsing image name %q: %v", imgName, err)
+			}
+		} else if srcRef != nil {
+			pullRef = srcRef
+		}
+		if pullRef.DockerReference() != nil {
+			reference = pullRef.DockerReference().String()
 		}
+
 		destRef, err := is.Transport.ParseStoreReference(r.store, reference)
 		if err != nil {
 			return errors.Errorf("error parsing dest reference name: %v", err)
 		}
-		if err = cp.Image(policyContext, destRef, srcRef, copyOptions); err != nil {
-			return errors.Errorf("error loading image %q: %v", image, err)
+		if err = cp.Image(policyContext, destRef, pullRef, copyOptions); err != nil {
+			fmt.Println("Failed")
+		} else {
+			return nil
```

Comment:
Is `images` intended to be a list of sources to try in order, terminating at the first success, or a list of sources which should all be copied? This `return nil`, and the search behavior of sys registries, suggest terminating at the first success, but the `DockerArchive` case is intended to pull all of them.

It can’t really be both at the same time. (And, actually, `DockerArchive` never worked as intended because `srcRef` is was constant for all of the `images` values.)

So, overall, perhaps the “parser” part should build a `[]struct imageToPull{attempts []struct singlePullAttempt{storageReferenceName string, pullSource types.ImageReference}}` or something like that?

Reply:
`images` was meant to be a list of image names found in the manifest.json or layout.json, since DockerArchive could have more than one image in it. (You are right that it never worked as intended because srcRef was constatnt - if that is the case should we remove that support out completely and just use the first name found in the list?)

For the case of registries, if it finds the image in one of the default registries it should pull that and stop - so try till first success.

Reply:
Right, but the point is that the same variable can’t implement the two different behaviors at the same time.

Reply:
Fixed. @mtrmac PTAL

Reply:
AFAICS `pullStructs` is still terminating on first success, but `DockerArchive` seems to want it to copy all images.